omnigibson/systems/macro_particle_system.py

```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
+        # reference the pre-existing one
+        if not get_prim_at_path(prim_path):
+            omni.kit.commands.execute(
+                "CopyPrim",
+                path_from=cls._particle_object.prim_path,
+                path_to=prim_path,
+            )
+            # Apply RigidBodyAPI to it so it is subject to physics
+            prim = get_prim_at_path(prim_path)
+            UsdPhysics.RigidBodyAPI.Apply(prim)
+            # PhysxSchema.PhysxRigidBodyAPI.Apply(prim)
+        return CollisionVisualGeomPrim(prim_path=prim_path, name=name)
+
+    @classmethod
+    def set_particle_template_object(cls, obj):
+        # Run super method
+        super().set_particle_template_object(obj=obj)
+
+        # Make sure min and max scale are identical
+        assert np.all(cls.min_scale == cls.max_scale), \
+            "Min and max scale should be identical for MacroPhysicalParticleSystem!"
+
+        # Compute particle radius
+        vertices = np.array(cls._particle_object.get_attribute("points")) * cls.max_scale.reshape(1, 3)
+        cls._particle_offset, cls._particle_radius = trimesh.nsphere.minimum_nsphere(trimesh.Trimesh(vertices=vertices))
+
+    @classmethod
+    def _refresh_particles_view(cls):
+        """
+        Internal helper method to refresh the particles' rigid body view to grab state
+
+        Should be called every time sim.play() is called
+        """
+        og.sim.pi.update_simulation(elapsedStep=0, currentTime=og.sim.current_time)
+        with suppress_omni_log(channels=["omni.physx.tensors.plugin"]):
+            cls.particles_view = og.sim.physics_sim_view.create_rigid_body_view(pattern=f"{cls.prim_path}/particles/*")
+
+    @classmethod
+    def clear(cls):
+        # Run super method first
+        super().clear()
+
+        # Clear internal variables
+        cls.particles_view = None
+
+    @classmethod
+    def remove_particle(cls, name):
+        # Run super first
+        super().remove_particle(name=name)
+
+        # Refresh particles view
+        cls._refresh_particles_view()
+
+    @classmethod
+    def add_particle(cls, prim_path, idn=None, scale=None):
+        # Run super first
+        super().add_particle(prim_path=prim_path, idn=idn, scale=scale)
+
+        # Refresh particles view
+        cls._refresh_particles_view()
+
+    @classmethod
+    def get_particles_position_orientation(cls, local=False):
+        if cls.n_particles > 0:
+            tfs = cls.particles_view.get_transforms()
+            pos, ori = tfs[:, :3], tfs[:, 3:]
+            pos = pos + T.quat2mat(ori) @ cls._particle_offset
+        else:
+            pos, ori = np.array([]).reshape(0, 3), np.array([]).reshape(0, 4)
+        return pos, ori
+
+    @classmethod
+    def get_particle_position_orientation(cls, idx, local=False):
+        positions, orientations = cls.get_particles_position_orientation()
+        return (positions[idx], orientations[idx]) if cls.n_particles > 0 else (positions, orientations)
```

Comment:
hmm is this the expected behavior? Maybe should raise an exception instead?

Reply:
done